openupgrade_scripts/scripts/sale_loyalty/16.0.1.0/pre-migration.py

```diff
@@ -0,0 +1,297 @@
+        """
+        ALTER TABLE sale_order_line
+        ADD COLUMN IF NOT EXISTS coupon_id INT
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE sale_order_line
+        SET coupon_id = loyalty_card.id
+        FROM loyalty_card
+        WHERE sale_order_line.order_id = loyalty_card.sales_order_id
+        """,
+    )
+    # Fill reward_identifier_code
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE sale_order_line
+        ADD COLUMN IF NOT EXISTS reward_identifier_code CHAR
+        """,
+    )
+    reward_code = _generate_random_reward_code()
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE sale_order_line
+        SET reward_identifier_code = %s
+        WHERE reward_id IS NOT NULL
+        """,
+        (reward_code,),
+    )
+    # Fill points_cost field
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE sale_order_line
+        ADD COLUMN IF NOT EXISTS points_cost FLOAT
+        """,
+    )
+    openupgrade.logged_query(
+        openupgrade.logged_query(
+            env.cr,
+            """
+            UPDATE sale_order_line
+            SET point_cost = ABS(untaxed_amount_to_invoice)
+            WHERE reward_id IS NOT NULL
+            AND untaxed_amount_to_invoice < 0
+            """,
+        )
+    )
+
+
+def delete_sql_constraints(env):
+    # Delete constraints to recreate it
+    openupgrade.delete_sql_constraint_safely(
+        env, "sale_loyalty", "sale_order_coupon_points", "order_coupon_unique"
+    )
+
+
+def update_template_keys(env):
+    """Update template keys of the merged sale_gift_card module in loyalty_sale"""
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE ir_ui_view
+        SET key = 'sale_loyalty.used_gift_card'
+        WHERE key = 'sale_gif_card.used_gift_card'
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE ir_ui_view
+        SET key = 'sale_loyalty.sale_purchased_gift_card'
+        WHERE key = 'sale_gif_card.sale_purchased_gift_card'
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE ir_ui_view
+        SET key = 'sale_loyalty.sale_orde_portal_content_inherit'
+        WHERE key = 'sale_gif_card.sale_orde_portal_content_inherit'
+        """,
+    )
+
+
+def merge_sale_gift_card_to_sale_loyalty_card(env):
+    # Update the coupon_id column in the sale_order_line table with the ID of the
+    # loyalty_card table based on certain criteria and relationships established between
+    # the loyalty_card, loyalty_reward and gift_card tables
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE sale_order_line AS sol
+        SET coupon_id = lc.id
+        FROM loyalty_card AS lc
+        JOIN loyalty_reward AS lr ON lc.program_id = lr.program_id
+        JOIN gift_card AS gc ON lc.program_id = gc.program_id
```

Comment:
Gift card model doesn't have program_id column, can you check what module add these field?

https://github.com/odoo/odoo/blob/15.0/addons/gift_card/models/gift_card.py

https://github.com/OCA/OCB/blob/15.0/addons/gift_card/models/gift_card.py

Reply:
This field is not added in any module, it is added in the migration scripts of the loyalty module in order to migrate the gift card data:

https://github.com/OCA/OpenUpgrade/blob/5912a50fa93234fe37643c1091919f780d718dd1/openupgrade_scripts/scripts/loyalty/16.0.1.0/pre-migration.py#L615-L621

I have indicated this in a comment to avoid confusion. Thanks for the corrections.